### PR TITLE
chore(server): core IoC #88 - createObjectFactory

### PR DESCRIPTION
**Pull request by fabis94**
## Checklist:

- [ ] My pull request follows the guidelines in the [Contributing guide](https://github.com/specklesystems/speckle-server/blob/main/CONTRIBUTING.md)?
- [ ] My pull request does not duplicate any other open [Pull Requests](../../pulls) for the same update/change?
- [ ] My commits are related to the pull request and do not amend unrelated code or documentation.
- [ ] My code follows a similar style to existing code.
- [ ] I have added appropriate tests.
- [ ] I have updated or added relevant documentation.

### CODE REVIEW

File: packages/server/modules/core/services/objects.js [hunk 1]
```diff
@@ -1,68 +1,15 @@
...
-          minDepth: object.__closure[prop]
-        })
-
-        if (totalChildrenCountByDepth[object.__closure[prop].toString()])
-          totalChildrenCountByDepth[object.__closure[prop].toString()]++
-        else totalChildrenCountByDepth[object.__closure[prop].toString()] = 1
-      }
-    }
-
-    delete insertionObject.__tree
```
Comment (fabis94) [inline packages/server/modules/core/services/objects.js:43]: @didimitrie @gjedlicska removed these 2 lines cause there was no way for these props to ever be there - prepInsertionObject doesn't return them

File: packages/server/modules/core/services/objects/management.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    data: stringifiedObj as unknown as Record<string, unknown>,
+    streamId,
+    id: obj.id,
+    // YEAH, this has been broken forever...
+    // speckleType: obj.speckleType
+    speckleType: obj.speckle_type
+  }
+}
+
+export const createObjectFactory =
```
Comment (fabis94) [inline packages/server/modules/core/services/objects/management.ts:51]: here's the new factory

File: packages/server/modules/core/services/objects/management.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  const objectByteSize = estimateStringMegabyteSize(stringifiedObj)
+  if (objectByteSize > MAX_OBJECT_SIZE_MB) {
+    throw new ObjectHandlingError(
+      `Object too large. Object ID: ${obj.id}. (${objectByteSize} MB is > than limit, ${MAX_OBJECT_SIZE_MB} MB)`
+    )
+  }
+
+  return {
+    // Not sure why we stringify it, cause knex will do that, but leaving it as is for now
+    data: stringifiedObj as unknown as Record<string, unknown>,
```
Comment (fabis94) [inline packages/server/modules/core/services/objects/management.ts:42]: @didimitrie @gjedlicska Any clues on why we do this (see comment)? Usually for Jsonb columns the TS/JS data type is an object, cause Knex knows how to handle the objects and convert them to serialized JSON when inserting, and parsing the JSON when retrieving from the DB.

Is this some kind of preemptive optimization or?

Comment (fabis94) [reply packages/server/modules/core/services/objects/management.ts:42]: Same goes for totalChildrenCountByDepth below

Comment (fabis94) [reply packages/server/modules/core/services/objects/management.ts]: i'm seeing this used for obj size calculations later on, so i've just adjusted the insertion type to support strings

File: packages/server/modules/core/domain/objects/types.ts [hunk 4]
```diff
@@ -1,3 +1,26 @@
...
+import { OverrideProperties, SetOptional } from 'type-fest'
+
+export type SpeckleObjectClosureEntry = ObjectChildrenClosureRecord
 
 export type SpeckleObject = ObjectRecord
+
+/**
+ * We preemptively serialize a couple of fields (usually knex does it)
+ */
+export type InsertableSpeckleObject = OverrideProperties<
```
Comment (fabis94) [inline packages/server/modules/core/domain/objects/types.ts:12]: @didimitrie @gjedlicska makes sense?

File: packages/server/modules/core/domain/objects/types.ts [hunk 5]
```diff
@@ -1,3 +1,26 @@
...
+  SetOptional<SpeckleObject, 'createdAt'>,
+  {
+    data: string
+    totalChildrenCountByDepth: string
+  }
+>
+
+export type RawSpeckleObject = Record<string, unknown> & {
```
Comment (fabis94) [inline packages/server/modules/core/domain/objects/types.ts:20]: @didimitrie @gjedlicska Does this make sense for the type of the raw object sent into the API (not the final object DB record)? I'm not sure about 'totalChildrenCount' and '__closure' - maybe they should be optional? I was thrown off by a `__closure !== null` check which made me think `undefined` is not a supported value there

Comment (fabis94) [reply packages/server/modules/core/domain/objects/types.ts:20]: I think `id` could be optional as well, it gets generated by prepInsertionObject. speckle_type seems to be the only actually required prop

Comment (fabis94) [reply packages/server/modules/core/domain/objects/types.ts:20]: See final versions of InsertableSpeckleObject & RawSPeckleObject here - https://github.com/specklesystems/speckle-server/pull/3335/files#diff-d35068824944d012d2120242ee1ed42bfc11ca8d6a257455e3039c48a3f0776d

**Review by gjedlicska — APPROVED** [review]

**Pull request merged**